docs/rfcs/xxxx-backwards-compatibility.md

```diff
@@ -0,0 +1,111 @@
+
+In the context of semantic versioning, a 1.0 release indicates that future minor releases have backwards compatibility with older minor releases in the same major release; version 1.1 is backwards compatible with version 1.0. Having major and minor releases signals to users when an upgrade may take more time (major releases) and when they can upgrade with more confidence (minor releases).
+
+However, Grafana Agent is a large project, with a large surface area for what may be considered part of the backwards compatibility guarantees. This proposal formally establishes what parts of Grafana Agent will be protected by backwards compatibility. 
+
+## Goals 
+
+- Set expectations for what is covered by backwards compatibility. 
+- Set expectations for when upgrades to new major versions will be forced.
+
+## Proposal 
+
+Backwards compatibility means that a user can upgrade their version of Grafana Agent without needing to make any changes with the way they interact with Grafana Agent, provided that interaction is within scope of being covered by backwards compatibility.
+
+### Scope of backwards compatibility   
+
+The following will be protected by backwards compatibility between minor releases: 
+
+- **User configuration**, including configuration file syntax and the command-line interface. 
+
+- **Versioned network APIs**, if any versioned APIs are introduced prior to the 1.0 release.
+
+- **Telemetry data used in official dashboards**. This means that users will continue to be able to use the same set of dashboards we provide when upgrading minor releases.    
+
+  - Official dashboards are dashboards in the repository's `operations/` [directory](../../operations/).
+
+- **Externally importable Go packages**. If a user is importing our code as a dependency, they should be able to upgrade to a new minor release without having to make changes to their code.
+
+  This means that names, signatures, and meanings may not change; a function called Calculate may not be changed from returning the sum of two numbers to returning the difference of two numbers.
+
+  Protecting Go packages with backwards compatibility is expected to be a contentious proposal. However, it has positive benefits: 
+
+  - It makes us good open-source citizens, allowing people to use Grafana Agent as a library with the same upgrade confidence as people who use Grafana Agent as a binary.  
+
+  - It allows us to get in the habit of being deliberate around when something should be consumable downstream, and will help us strengthen our skills with designing good APIs in general. Ideally, this helps us get better at writing Go code _and_ writing Flow components.
+
+  - It will help train us to more easily recognize when any kind of change is breaking. 
+
+  The easiest way to commit to this without any effort is to put everything in an top-level internal folder, effectively saying that Grafana Agent has no code API that can be consumed downstream.
```

Comment:
Not a fan of this easy step. Everything to internal feels less open then breaking having api changes. If someone is using grafana agent as a library and things they need are behind internal then that requires a fork instead of changing their api signatures.

Reply:
> Not a fan of this easy step. Everything to internal feels less open then breaking having api changes. If someone is using grafana agent as a library and things they need are behind internal then that requires a fork instead of changing their api signatures.

That's exactly the intent; having something be exposed should be done with intent that we're ready for it to be consumed publicly and it's been designed in a way that's broadly useful. 

Forking instead of importing isn't always bad, as [A little copying is better than a little dependency](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=9m28s).